packages/backend/src/apps/m365-excel/actions/write-cell-values/index.ts

```diff
@@ -0,0 +1,119 @@
+import type { IRawAction } from '@plumber/types'
+
+import { generateStepError } from '@/helpers/generate-step-error'
+
+import WorkbookSession from '../../common/workbook-session'
+
+import { parametersSchema } from './parameters-schema'
+
+const action: IRawAction = {
+  name: 'Write cell values',
+  key: 'writeCellValues',
+  description: "Write values into a spreadsheet's cells",
+  arguments: [
+    {
+      key: 'fileId',
+      label: 'Excel File',
+      required: true,
+      description: 'File to edit',
+      type: 'dropdown' as const,
+      variables: false,
+      source: {
+        type: 'query' as const,
+        name: 'getDynamicData' as const,
+        arguments: [
+          {
+            name: 'key',
+            value: 'listFiles',
+          },
+        ],
+      },
+    },
+    {
+      key: 'worksheetId',
+      label: 'Worksheet',
+      required: true,
+      description: 'Worksheet to edit',
+      type: 'dropdown' as const,
+      variables: false,
+      source: {
+        type: 'query' as const,
+        name: 'getDynamicData' as const,
+        arguments: [
+          {
+            name: 'key',
+            value: 'listWorksheets',
+          },
+          {
+            name: 'parameters.fileId',
+            value: '{parameters.fileId}',
+          },
+        ],
+      },
+    },
+    {
+      label: 'Values',
+      key: 'cells',
+      type: 'multirow' as const,
+      required: true,
+      // We need to make 1 separate request for each cell, so limit to 3 as a
+      // balance between convenience and API quota usage.
+      description:
+        'Specify cells you want to write values to (max 3). Leave a value blank to clear the cell.',
+      subFields: [
+        {
+          label: 'Cell Address (e.g. A1)',
+          key: 'address',
+          type: 'string' as const,
+          required: true,
+          variables: true,
+        },
+        {
+          label: 'Value',
+          key: 'value',
+          type: 'string' as const,
+          required: false,
+          variables: true,
+        },
+      ],
+    },
+  ],
+
+  async run($) {
+    const parametersParseResult = parametersSchema.safeParse($.step.parameters)
+
+    if (parametersParseResult.success === false) {
```

Comment:
For some reason, typescript doesn't refine the type properly if I do `if (!parametersParseResult.success)`